internal/resource/informer.go

```diff
@@ -0,0 +1,271 @@
+func (ri *resourceInformer) Name() string {
+	return "resource-informer"
+}
+
+func (ri *resourceInformer) Init() error {
+	// ensure we can access procfs
+	_, err := ri.fs.AllProcs()
+	if err != nil {
+		return fmt.Errorf("failed to access procfs: %w", err)
+	}
+
+	ri.logger.Info("Resource informer initialized successfully")
+	return nil
+}
+
+func (ri *resourceInformer) Refresh() error {
+	started := ri.clock.Now()
+
+	procs, err := ri.fs.AllProcs()
+	if err != nil {
+		return fmt.Errorf("failed to get processes: %w", err)
+	}
+
+	// construct current running processes and containers
+	procsRunning := make(map[int]*Process, len(procs))
+	containersRunning := make(map[string]*Container)
+
+	// Refresh process cache and update running processes
+	var refreshErrs error
+	for _, p := range procs {
+		pid := p.PID()
+		// start by updating the process
+		proc, err := ri.updateProcessCache(p)
+		if err != nil {
+			ri.logger.Debug("Failed to get process info", "pid", pid, "error", err)
+			refreshErrs = errors.Join(refreshErrs, err)
+			continue
+		}
+		procsRunning[pid] = proc
+
+		if c := proc.Container; c != nil {
+			//  Containers: group processes by container
+			_, seen := containersRunning[c.ID]
+			containersRunning[c.ID] = ri.updateContainerCache(proc, seen)
+		}
+	}
+
+	// Find terminated processes
+	nodeCPUDelta := float64(0)
+	procsTerminated := make(map[int]*Process)
+	for pid, proc := range ri.procCache {
+		if _, isRunning := procsRunning[pid]; isRunning {
+			nodeCPUDelta += proc.CPUTimeDelta
+			continue
+		}
+		procsTerminated[pid] = proc
+		delete(ri.procCache, pid)
+	}
+
+	// Find terminated containers
+	totalContainerDelta := float64(0)
+	containersTerminated := make(map[string]*Container)
+	for id, container := range ri.containerCache {
+		if _, isRunning := containersRunning[id]; isRunning {
+			totalContainerDelta += container.CPUTimeDelta
+			continue
+		}
+		containersTerminated[id] = container
+		delete(ri.containerCache, id)
+	}
+
+	// Update tracking structures
+	ri.processes.NodeCPUTimeDelta = nodeCPUDelta
+	ri.processes.Running = procsRunning
+	ri.processes.Terminated = procsTerminated
+
+	ri.containers.NodeCPUTimeDelta = nodeCPUDelta
+	ri.containers.Running = containersRunning
+	ri.containers.Terminated = containersTerminated
+
+	now := ri.clock.Now()
+	ri.lastScanTime = now
+	duration := ri.clock.Since(started)
+
+	ri.logger.Debug("Resource information collected",
+		"process.running", len(procsRunning),
+		"process.terminated", len(procsTerminated),
+		"container.running", len(containersRunning),
+		"container.terminated", len(containersTerminated),
+		"duration", duration)
+
+	return refreshErrs
+}
+
+func (ri *resourceInformer) updateContainerCache(proc *Process, visited bool) *Container {
+	c := proc.Container
+	if c == nil {
+		return nil
+	}
+
+	cached, exists := ri.containerCache[c.ID]
+	if !exists {
+		cached = c.Clone()
+		ri.containerCache[c.ID] = cached
+	}
+
+	if !visited {
+		cached.CPUTimeDelta = 0
+	}
```

Comment:
`visited` means that the container is already added to `containersRunning` during the same monitor interval, and we want to reset the CPUTimeDelta for the cached Container if we are seeing this container for the first time in the interval.

it this understanding correct? can we reset the cputime explicitly at the end of interval?

Reply:
Do you mean have timer that resets it ? If thats what you meant, then we will have to think of concurrency :(

Reply:
no, at the end of monitor interval loop.